### PR TITLE
add check that IR type and expression type are the same

### PR DESCRIPTION
**Pull request by catoverdrive**
This caught a bug in the type inference in `GroupByKey`, which is also fixed in the PR.

@konradjk I'm pretty sure that this makes #5147 work, but I haven't confirmed directly so if you could confirm that it does and then close (or re-ping) once this goes in, that would be great.

### CODE REVIEW

**Comment by konradjk** [comment]
Appears to work! Fixes #5147

File: hail/python/test/hail/expr/test_types.py [hunk 1]
```diff
@@ -82,3 +82,5 @@ def test_coercers_can_coerce(self):
             c = coercer_from_dtype(t)
             self.assertTrue(c.can_coerce(t))
             self.assertFalse(c.requires_conversion(t))
+
+
```
Comment (chrisvittal) [inline hail/python/test/hail/expr/test_types.py]: I noticed that you removed a test. I would like to see a regression test for this.

Comment (catoverdrive) [reply hail/python/test/hail/expr/test_types.py]: I didn't remove it, actually. I had added a test to try to reproduce the bug, realized it didn't actually reproduce the bug, and took it back out. I'll remove the whitespace.

**Comment by catoverdrive** [comment]
This bug is caught in expr.test_array_methods, on the `group_by` line with the additions I made in this PR; I think any test that I would add would essentially just be a duplicate of that.

File: hail/python/test/hail/expr/test_types.py (hunk 1 above)
Comment (catoverdrive) [reply hail/python/test/hail/expr/test_types.py]: ah, I see. I thought you'd meant "removed an existing test from the test suite", but you probably just meant "removed the test that I'd added in a previous commit". But yeah, I'm pretty sure the regression test already exists in `test_array_methods`, and the test I'd added turned out to not actually reproduce the bug.

Comment (chrisvittal) [reply hail/python/test/hail/expr/test_types.py]: Alright. Good to know.

**Pull request merged**